### PR TITLE
DONOTMERGE: Review of azure-identity

### PR DESCRIPTION
**Pull request by chlowell**

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#
+# NOTE: This dynamically typed stub was automatically generated by stubgen.
+
+from .._authn_client import AuthnClient
+from .._base import CertificateCredentialBase, ClientSecretCredentialBase
+from azure.core.credentials import AccessToken
+from typing import Any, Mapping
+
+class ClientSecretCredential(ClientSecretCredentialBase):
+    def __init__(self, client_id: str, secret: str, tenant_id: str, **kwargs: Mapping[str, Any]) -> None: ...
```
Comment (KieranBrantnerMagee) [inline sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi]: Sometimes we do client_id, tenant_id, <otherstuff including secrets> and sometimes we put secret between client/tenant.  I prefer the former, but regardless consistency would seem desirable.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/environment.pyi [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+from .._constants import EnvironmentVariables
+from .client_credential import CertificateCredential, ClientSecretCredential
+from .user import UsernamePasswordCredential
+from azure.core.credentials import AccessToken
+from typing import Any, Mapping
+
+EnvironmentCredentialTypes: Any
+
+class EnvironmentCredential:
+    def __init__(self, **kwargs: Mapping[str, Any]) -> None: ...
```
Comment (KieranBrantnerMagee) [inline sdk/identity/azure-identity/out/azure/identity/_credentials/environment.pyi]: Is there a reason our kwargs types vary so much?  (Mapping[] vs Any)  If it matters at all, I'd bang the consistency hammer again, but if this is premature ignore moe.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Stubs for azure.identity._credentials.user (Python 3)
+#
+# NOTE: This dynamically typed stub was automatically generated by stubgen.
+
+from .._authn_client import AuthnClient, AuthnClientBase
+from .._internal import PublicClientCredential, wrap_exceptions
+from azure.core.credentials import AccessToken
+from typing import Any, Callable, Optional
+
+class DeviceCodeCredential(PublicClientCredential):
```
[the remark this answers is not in the document]
Comment (chlowell) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi]: Yes, that's the idea. `InteractiveBrowserCredential` will block and (attempt to) browse to a page where the user needs to provide input. `DeviceCodeCredential` is similar but the browsing and input is manual.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/environment.pyi (hunk 2 above)
Comment (chlowell) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/environment.pyi]: Accidents of history, I'll make them consistently `Any` when I do a mypy pass.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi (hunk 1 above)
Comment (johanste) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi]: We should do `tenant_id, client_id` whenever they are required. If they are not required, they should be `**kwargs` and not positional.

File: sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#
+# NOTE: This dynamically typed stub was automatically generated by stubgen.
+
+from ..._base import CertificateCredentialBase, ClientSecretCredentialBase
+from .._authn_client import AsyncAuthnClient
+from azure.core.credentials import AccessToken
+from typing import Any, Mapping
+
+class ClientSecretCredential(ClientSecretCredentialBase):
+    def __init__(self, client_id: str, secret: str, tenant_id: str, **kwargs: Mapping[str, Any]) -> None: ...
```
Comment (johanste) [inline sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi]: Change parameter order - `self, tenant_id, client_id, secret, **kwargs`

File: sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    async def get_token(self, *scopes: str, **kwargs: Any) -> AccessToken: ...
+
+class CertificateCredential(CertificateCredentialBase):
+    def __init__(self, client_id: str, tenant_id: str, certificate_path: str, **kwargs: Mapping[str, Any]) -> None: ...
```
Comment (johanste) [inline sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi]: Change parameter order - `self, tenant_id, client_id, secret, **kwargs`

File: sdk/identity/azure-identity/out/azure/identity/_credentials/browser.pyi [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Stubs for azure.identity._credentials.browser (Python 3)
+#
+# NOTE: This dynamically typed stub was automatically generated by stubgen.
+
+from .._internal import AuthCodeRedirectServer, PublicClientCredential, wrap_exceptions
+from azure.core.credentials import AccessToken
+from typing import Any
+
+class InteractiveBrowserCredential(PublicClientCredential):
+    def __init__(self, client_id: str, **kwargs: Any) -> None: ...
```
Comment (schaabs) [inline sdk/identity/azure-identity/out/azure/identity/_credentials/browser.pyi]: client_id should be optional here and moved to kwargs

File: sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi (hunk 1 above)
Comment (schaabs) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/client_credential.pyi]: Agreed. https://github.com/Azure/azure-sdk-for-python/issues/7951

File: sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi (hunk 3 above)
[the remark this answers is not in the document]
Comment (schaabs) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi]: My only objection to this is that InteractiveBrowserCredential we do the interactive flow automatically for the user, whereas in this case the developer must provide a callback for the interaction. Thoughts?

File: sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi (hunk 4 above)
Comment (schaabs) [reply sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi]: https://github.com/Azure/azure-sdk-for-python/issues/7951

File: sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi (hunk 5 above)
Comment (schaabs) [reply sdk/identity/azure-identity/out/azure/identity/aio/_credentials/client_credential.pyi]: https://github.com/Azure/azure-sdk-for-python/issues/7951

File: sdk/identity/azure-identity/out/azure/identity/aio/_credentials/managed_identity.pyi [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+from ..._constants import Endpoints, EnvironmentVariables
+from .._authn_client import AsyncAuthnClient
+from azure.core import Configuration
+from azure.core.credentials import AccessToken
+from azure.identity._credentials.managed_identity import _ManagedIdentityBase
+from typing import Any, Optional
+
+class ManagedIdentityCredential:
+    def __new__(cls, *args: Any, **kwargs: Any): ...
+    def __init__(self, client_id: Optional[str] = ..., **kwargs: Any) -> None: ...
```
Comment (schaabs) [inline sdk/identity/azure-identity/out/azure/identity/aio/_credentials/managed_identity.pyi]: client_id probably should be in kwargs here.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi (hunk 3 above)
[the remark this answers is not in the document]
Comment (chlowell) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi]: The callback is optional. If not provided, we print the device code and auth URL.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/authorization_code.pyi [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+# NOTE: This dynamically typed stub was automatically generated by stubgen.
+
+from .._internal.aad_client import AadClient
+from azure.core.credentials import AccessToken
+from typing import Any, Optional
+
+class AuthorizationCodeCredential:
+    def __init__(
+        self,
+        client_id: str,
```
Comment (johanste) [inline sdk/identity/azure-identity/out/azure/identity/_credentials/authorization_code.pyi]: `tenant_id, client_id`

Comment (johanste) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/authorization_code.pyi]: #7951

File: sdk/identity/azure-identity/out/azure/identity/_credentials/browser.pyi (hunk 6 above)
Comment (chlowell) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/browser.pyi]: Defaulting to Azure CLI?

File: sdk/identity/azure-identity/out/azure/identity/_credentials/managed_identity.pyi [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    def __new__(cls, *args: Any, **kwargs: Any): ...
+    def __init__(self, client_id: Optional[str] = ..., **kwargs: Any) -> None: ...
+    def get_token(self, *scopes: str, **kwargs: Any) -> AccessToken: ...
+
+class _ManagedIdentityBase:
+    def __init__(
+        self,
+        endpoint: str,
+        client_cls: Type,
+        config: Optional[Configuration] = ...,
```
Comment (johanste) [inline sdk/identity/azure-identity/out/azure/identity/_credentials/managed_identity.pyi]: I need to look at the `config` option.

File: sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi (hunk 3 above)
[the remark this answers is not in the document]
Comment (schaabs) [reply sdk/identity/azure-identity/out/azure/identity/_credentials/user.pyi]: no changes

**Comment by chlowell** [comment]
Addressed by #7980

**Pull request closed without merging**